### PR TITLE
adding updated scripts to fetch the data and convert it into json

### LINKED ISSUE #1: update Makefile and scrape_table.py to pull from updated data source
current Makefile produces the following error:

```bash
Traceback (most recent call last):
  File "/home/user/Documents/chicago-elections-archive/scripts/scrape_table.py", line 69, in <module>
    main()
  File "/home/user/Documents/chicago-elections-archive/scripts/scrape_table.py", line 56, in main
    candidates = get_candidates(soup)
  File "/home/user/Documents/chicago-elections-archive/scripts/scrape_table.py", line 9, in get_candidates
    headers = [v.get_text().strip() for v in soup.find("table").find_all("b")]
AttributeError: 'NoneType' object has no attribute 'find_all'
```

Looks like the chicagoelections.gov has updated their pages that produce these results. https://chicagoelections.gov/en/election-results-specifics.asp is now a 404 page. This is the new one https://chicagoelections.gov/elections/results/

### PR DESCRIPTION
**Pull request by yashBhosale**
Working on a new script to fetch all of the election data. This is not the finished script but it is at 90%. Finishing touches needed. Will close #1 when complete.

### CODE REVIEW

**Comment by yashBhosale** [comment]
This code could likely be pared down a bit - i was mostly putting the tracks down on as I was driving over them. Ideally I can finish this in the morning and we can merge this.

File: scripts/scrape_elections.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    return subtables
+
+async def main():
+
+    results_metadata: dict = load(open("../output/results-metadata.json", "r"))
+    pairs = [(contest, race) for contest, c_info in results_metadata.items() for race in c_info["races"]]
+    #print(len(pairs))
+    #async with ClientSession() as cs:
+    #    async with cs.get("https://chicagoelections.gov/elections/results/156/download?contest=15&ward=&precinct=") as resp:
```
Comment (yashBhosale) [inline scripts/scrape_elections.py:51]:
```
def fetch(race: int, contest: int, cs: ClientSession):
 
   resp = await cs.get("https://chicagoelections.gov/elections/results/{race}/download?contest={content}&ward=&precinct=")
   return book_pandas(await resp.content.read(), race, contest)
 
```
We can `asyncio.gather` this and it should reduce time wasted on I/o which is likely the biggest performance killer.

File: scripts/scrape_elections.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import xlrd
+from pprint import pprint
+import pandas as pd
+from aiohttp import ClientSession
+from json import load, dump
+from asyncio import run
+
+
+
+def book_pandas(book: BytesIO):
```
Comment (yashBhosale) [inline scripts/scrape_elections.py:11]: def book_pandas(book, race, contest):

File: scripts/scrape_elections.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            pass
+        cols = sub_table[0]
+        print(sub_table)
+        print(cols)
+        for i in range(len(cols)):
+            if cols[i] == "%":
+                cols[i] = f"{cols[i-1]} %"
+        subtables[ward] = pd.DataFrame(sub_table[1:], columns=sub_table[0]).set_index('Precinct').to_dict(orient="index")
+        cur_row = next(rows, None)
+    dump(subtables, open("subtable.json", 'w'))
```
Comment (yashBhosale) [inline scripts/scrape_elections.py:41]: `f"{race}_{contest}.json" or something. Oh and probably pickle for prod.

File: scripts/scrape_elections.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        cur_row = next(rows)
+        sub_table = []
+        try:
+            while not all([cell.ctype in (xlrd.XL_CELL_EMPTY, xlrd.XL_CELL_BLANK) for cell in cur_row]):
+                sub_table.append([cell.value for cell in cur_row])
+                cur_row = next(rows)
+        except StopIteration:
+            pass
+        cols = sub_table[0]
+        print(sub_table)
```
Comment (yashBhosale) [inline scripts/scrape_elections.py:34]: Remove

File: scripts/scrape_elections.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        print(cols)
+        for i in range(len(cols)):
```
Comment (yashBhosale) [inline scripts/scrape_elections.py:36]: 'cols = [col if col != '%' else cols[i-1] + " %" for i, cols in enumerate(cols)]'

File: scripts/scrape_elections.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
+from io import BytesIO
+import xlrd
+from pprint import pprint
+import pandas as pd
+from aiohttp import ClientSession
+from json import load, dump
+from asyncio import run
+
+
```
Comment (yashBhosale) [inline scripts/scrape_elections.py:9]: We need to check for edge cases.

File: scripts/scrape_elections.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+async def main():
+
+    results_metadata: dict = load(open("../output/results-metadata.json", "r"))
+    pairs = [(contest, race) for contest, c_info in results_metadata.items() for race in c_info["races"]]
+    #print(len(pairs))
+    #async with ClientSession() as cs:
+    #    async with cs.get("https://chicagoelections.gov/elections/results/156/download?contest=15&ward=&precinct=") as resp:
+    #        book_pandas(await resp.content.read())
+    book_pandas(open("/home/yash/Downloads/download.xls", "rb").read())
```
Comment (yashBhosale) [inline scripts/scrape_elections.py:53]: Oops. Remove.

File: scripts/scrape_elections.py (hunk 1 above)
Comment (yashBhosale) [reply scripts/scrape_elections.py:51]: also need to pool this

File: scripts/scrape_elections.py (hunk 2 above)
Comment (derekeder) [reply scripts/scrape_elections.py:11]: is this a todo note for later?

File: scripts/scrape_elections.py (hunk 1 above)
Comment (derekeder) [reply scripts/scrape_elections.py:51]: we should also cache the request response. these will likely never change

**Review by derekeder — COMMENTED** [review]
I did an initial review. I think for this to be ready to bring in we'll want to have it:

- save the results in the same place as the old script
- remove the unused scraper code that this replaces
- update the readme as necessary 

We can do it here, or in a future PR, but the `elections.json` file seems like something we could scrape and generate dynamically based on the HTML on this page: https://chicagoelections.gov/elections/results

File: scripts/scrape_elections.py (hunk 2 above)
Comment (yashBhosale) [reply scripts/scrape_elections.py:11]: Yeah

**Comment by yashBhosale** [comment]
Never mind, I did figure it out. It's trivial enough that I'll do it in the morning and hopefully this should be good enough to merge.

**Comment by derekeder** [comment]
@yashBhosale sounds good! we can plan to look at it tonight

**Pull request merged**